File: leetcode-solutions/arrays_and_hashing/productOfArrayExceptSelf.js
const nums = [-1, 1, 0, -3, 3];
// const nums = [1, 2, 3, 4];

const productExceptSelf = function (nums) {
  let output = [];
  for (let i = 0; i < nums.length; i++) {
    let trim = nums.slice(0, i).concat(nums.slice(i + 1));
    output.push(trim);
  }
  return output.map((el) => el.reduce((a, b) => a * b));
};

console.log(productExceptSelf(nums));

File: leetcode-solutions/arrays_and_hashing/productOfArrayExceptSelf.py
from functools import reduce

nums = [1, 2, 3, 4]


# def productOfArrayExceptSelf(nums):
#     output = []
#     for i in range(len(nums)):
#         trim = nums[:i] + nums[i + 1 :]
#         result = reduce(lambda x, y: x * y, trim)
#         output.append(result)
#     return output


def productOfArrayExceptSelf(nums):
    output = []
    pre = 1
    post = 1
    for i in range(len(nums)):
        if i == 0:
            output.append(pre)
        else:
            pre *= nums[i - 1]
            output.append(pre)
    for i in reversed(range(len(nums))):
        if i == len(nums) - 1:
            output[i] *= 1
        else:
            post *= nums[i + 1]
            output[i] *= post
    return output


print(productOfArrayExceptSelf(nums))
